src/app/clusters/color-control-server/color-control-server.cpp

```diff
@@ -47,711 +42,1080 @@
-
-#define MIN_CURRENT_LEVEL 0x01
-#define MAX_CURRENT_LEVEL 0xFE
-
-#define REPORT_FAILED 0xFF
-
-typedef struct
-{
-    uint8_t initialHue;
-    uint8_t currentHue;
-    uint8_t finalHue;
-    uint16_t stepsRemaining;
-    uint16_t stepsTotal;
-    uint16_t initialEnhancedHue;
-    uint16_t currentEnhancedHue;
-    uint16_t finalEnhancedHue;
-    EndpointId endpoint;
-    bool up;
-    bool repeat;
-    bool isEnhancedHue;
-} ColorHueTransitionState;
-
-static ColorHueTransitionState colorHueTransitionState;
-
-typedef struct
-{
-    uint16_t initialValue;
-    uint16_t currentValue;
-    uint16_t finalValue;
-    uint16_t stepsRemaining;
-    uint16_t stepsTotal;
-    uint16_t lowLimit;
-    uint16_t highLimit;
-    EndpointId endpoint;
-} Color16uTransitionState;
-
-static Color16uTransitionState colorXTransitionState;
-static Color16uTransitionState colorYTransitionState;
-
-static Color16uTransitionState colorTempTransitionState;
-
-static Color16uTransitionState colorSaturationTransitionState;
-
-// Forward declarations:
-static bool computeNewColor16uValue(Color16uTransitionState * p);
-static void stopAllColorTransitions(void);
-static void handleModeSwitch(EndpointId endpoint, uint8_t newColorMode);
-static bool shouldExecuteIfOff(EndpointId endpoint, uint8_t optionMask, uint8_t optionOverride);
+using namespace app::Clusters;
+
+/**********************************************************
+ * Attributes Definition
+ *********************************************************/
+
+ColorControlServer ColorControlServer::instance;
 
 #ifdef EMBER_AF_PLUGIN_COLOR_CONTROL_SERVER_HSV
-static uint8_t addHue(uint8_t hue1, uint8_t hue2);
-static uint8_t subtractHue(uint8_t hue1, uint8_t hue2);
-static uint8_t addSaturation(uint8_t saturation1, uint8_t saturation2);
-static uint8_t subtractSaturation(uint8_t saturation1, uint8_t saturation2);
-static void initHueSat(EndpointId endpoint);
-static uint8_t readHue(EndpointId endpoint);
-static uint8_t readSaturation(EndpointId endpoint);
-static uint16_t addEnhancedHue(uint16_t hue1, uint16_t hue2);
-static uint16_t subtractEnhancedHue(uint16_t hue1, uint16_t hue2);
-static uint16_t readEnhancedHue(EndpointId endpoint);
-
-static bool moveHue(uint8_t moveMode, uint16_t rate, uint8_t optionsMask, uint8_t optionsOverride, bool isEnhanced);
-static bool moveToHue(uint16_t hue, uint8_t hueMoveMode, uint16_t transitionTime, uint8_t optionsMask, uint8_t optionsOverride,
-                      bool isEnhanced);
-static bool moveToHueAndSaturation(uint16_t hue, uint8_t saturation, uint16_t transitionTime, uint8_t optionsMask,
-                                   uint8_t optionsOverride, bool isEnhanced);
-static bool stepHue(uint8_t stepMode, uint16_t stepSize, uint16_t transitionTime, uint8_t optionsMask, uint8_t optionsOverride,
-                    bool isEnhanced);
+std::unordered_map<chip::EndpointId, ColorControlServer::ColorHueTransitionState> ColorControlServer::colorHueTransitionStates;
```

Comment:
We cannot/should not use STL containers that allocate from heap within cluster implementations that are expected to go onto constrained devices.

Reply:
In addition, these states that are private to the color control server should really be members within that class. If a single instance of this class resides on a single endpoint, you can then do away with having a map to store the different associations between endpoint to state.

The instantiation of this class on multiple endpoints can then be code-generated.

Reply:
@mrjerryjohns 

> In addition, these states that are private to the color control server should really be members within that class. If a single instance of this class resides on a single endpoint, you can then do away with having a map to store the different associations between endpoint to state.

For the static attributes transitionStates, they shouldn't be static since the color control server is a singleton. 

> We cannot/should not use STL containers that allocate from heap within cluster implementations that are expected to go onto constrained devices.

Since matter supports dynamic endpoints, i don't see how we can avoid using the heap to store endpoint spefic variables unless we reserve stack memory for the maximum number of endpoints that can exists at runtime which would result, in most cases, in a waste of memory.

> The instantiation of this class on multiple endpoints can then be code-generated.

Same applies to code generation since apps can create endpoints at runtime.